godtools/App/Services/Renderer/Views/MobileContent/Views/Pages/MobileContentPagesViewModel.swift

```diff
@@ -68,8 +64,379 @@ class MobileContentPagesViewModel: NSObject {
+            
+        case .visiblePages:
+            
+            pageModelsToRender = pageRenderer.getVisibleRenderablePageModels()
+        }
+        
+        rendererWillChangeSignal.accept()
+        
+        currentPageRenderer.send(pageRenderer)
+                
+        self.pageModels = pageModelsToRender
+        
+        numberOfPages.accept(value: pageModels.count)
+    }
+    
+    // MARK: - Page Life Cycle
+    
+    func pageWillAppear(page: Int) -> MobileContentView? {
+        
+        guard let window = self.window, let safeArea = self.safeArea else {
+            return nil
+        }
+        
+        guard page >= 0 && page < pageModels.count else {
+            return nil
+        }
+                
+        let renderPageResult: Result<MobileContentView, Error> =  currentPageRenderer.value.renderPageModel(
+            pageModel: pageModels[page],
+            page: page,
+            numberOfPages: pageModels.count,
+            window: window,
+            safeArea: safeArea,
+            trainingTipsEnabled: trainingTipsEnabled
+        )
+        
+        switch renderPageResult {
+        
+        case .success(let mobileContentView):
+            return mobileContentView
+            
+        case .failure( _):
+            break
+        }
+        
+        return nil
+    }
+    
+    func pageDidAppear(page: Int) {
+        
+        currentRenderedPageNumber = page
+        
+        if page > highestPageNumberViewed {
+            highestPageNumberViewed = page
+        }
+    }
+    
+    func pageDidDisappear(page: Int) {
+              
+        let didNavigateBack: Bool = currentRenderedPageNumber < page
+        let shouldRemoveAllFollowingPages: Bool = initialPageRenderingType == .chooseYourOwnAdventure && didNavigateBack
+        
+        if shouldRemoveAllFollowingPages {
+            removeFollowingPagesFromPage(page: currentRenderedPageNumber)
+        }
+        
+        removePageIfHidden(page: page)
+    }
+    
+    func didChangeMostVisiblePage(page: Int) {
+        
+        currentRenderedPageNumber = page
+    }
+}
+
+// MARK: - Page Navigation
+
+extension MobileContentPagesViewModel {
+    
+    private func navigateToPage(page: MobileContentPagesPage, forceReloadPagesUI: Bool, animated: Bool) {
+        
+        let isChooseYourOwnAdventure: Bool = initialPageRenderingType == .chooseYourOwnAdventure
+        let pageRenderer: MobileContentPageRenderer = currentPageRenderer.value
+        let renderablePages: [Page] = pageRenderer.getRenderablePageModels()
+        
+        let pagesToRender: [Page]?
+        let navigateToPageModel: Page?
+        let shouldReloadPagesUI: Bool
+        
+        switch page {
+            
+        case .pageId(let value):
+                     
+            if isChooseYourOwnAdventure {
+                
+                let introPageId: String = "intro"
+                let categoriesPageId: String = "categories"
```

Comment:
you shouldn't hardcode these, this is specific to the "openers" tool only and wouldn't be applicable to other cyoa tools. There is a `page.parentPage` property that will return the configured page that should be added to the back stack. You should be able to recurse through the hierarchy to build the back stack

Reply:
https://github.com/CruGlobal/kotlin-mpp-godtools-tool-parser/blob/main/module/parser/src/commonMain/kotlin/org/cru/godtools/shared/tool/parser/model/page/Page.kt#L112

Reply:
this overall logic could look something like:

```
if isChooseYourOwnAdventure {
   var pages = [renderablePages.first(where: {$0.id == value})]
   
   while {
      let parent = pages[0].parentPage
      if (parent == null || pages.contains(parent)) break
      pages = parent + pages
   }
   
   pagesToRender = pages
}
```

Reply:
Thanks @frett !

Reply:
Hey @frett this is ready for a second look.  I was able to replace the entire pageId navigation with the logic of moving up the parent.